events/roommessageevent.h

```diff
@@ -75,59 +80,71 @@ namespace QMatrixClient
                 QString body;
         };
 
-        class ImageContent: public Base
+        class FileInfo: public Base
         {
             public:
+                FileInfo(QUrl u, const QJsonObject& infoJson,
+                         QString originalFilename = QString());
+
                 QUrl url;
-                int height;
-                int width;
-                int size;
-                QString mimetype;
+                int fileSize;
+                QMimeType mimetype;
+                QString originalName;
         };
 
-        class FileContent: public Base
+        class ImageInfo: public FileInfo
         {
             public:
-                QString filename;
-                QString mimetype;
-                int size;
-                QUrl url;
+                ImageInfo(QUrl u, const QJsonObject& infoJson);
+
+                QSize imageSize;
         };
 
+        template <class ContentInfoT>
+        class ThumbnailedContent: public ContentInfoT
+        {
+            public:
+                ThumbnailedContent(const QJsonObject& json)
+                    : ContentInfoT(json["url"].toString(), json["info"].toObject())
+                    , thumbnail(json["thumbnail_url"].toString(),
+                                json["thumbnail_info"].toObject())
+                { }
+
+                ImageInfo thumbnail;
+        };
+
+        using ImageContent = ThumbnailedContent<ImageInfo>;
+        using FileContent = ThumbnailedContent<FileInfo>;
+
         class LocationContent: public Base
         {
             public:
+                LocationContent(const QJsonObject& json);
+
                 QString geoUri;
-                int thumbnailHeight;
-                int thumbnailWidth;
-                QString thumbnailMimetype;
-                int thumbnailSize;
-                QUrl thumbnailUrl;
+                ImageInfo thumbnail;
         };
 
-        class VideoContent: public Base
+        // The spec structures m.video messages differently for some reason -
```

Comment:
Good call, actually - it didn't come to my mind they could do it inadvertently. Asked in #matrix-dev.